### PR TITLE
GIT-6: Clone non-public repositories

### PR DESCRIPTION
**Pull request by Haxsen**
[Issue Link](https://jira.xwiki.org/projects/GIT/issues/GIT-6)

Glad to correct my mistakes.
Thanks for your reviews and valuable time :)

* Before:
PrivateGitBefore

* After:
PrivateGitAfter

### CODE REVIEW

File: src/main/java/org/xwiki/git/GitManager.java [hunk 1]
```diff
@@ -48,6 +48,24 @@
...
+     * Clone a Private Git repository using the credentials provided by user and store it locally in the
+     * XWiki Permanent directory. If the repository is already cloned, no action is done.
+     *
+     * @param repositoryURI the URI to the Git repository to clone (eg "git://github.com/xwiki/xwiki-commons.git")
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of Git user
+     * @param token the Personal access token generated by the user
+     * @return the cloned Repository instance
+     * @since 12.3RC1
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/GitManager.java]: This is not correct. I doubt that the version of the Git extension is at 12.3RC1 :)

File: src/main/java/org/xwiki/git/GitManager.java [hunk 2]
```diff
@@ -48,6 +48,24 @@
...
+     */
+    default Repository getPrivateRepository(String repositoryURI, String localDirectoryName,
+                                            String username, String token)
+    {
+        throw new UnsupportedOperationException();
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/GitManager.java:68]: Cool, you found it :)

File: src/main/java/org/xwiki/git/GitManager.java [hunk 3]
```diff
@@ -48,6 +48,24 @@
...
 
+    /**
+     * Clone a Private Git repository using the credentials provided by user and store it locally in the
+     * XWiki Permanent directory. If the repository is already cloned, no action is done.
+     *
+     * @param repositoryURI the URI to the Git repository to clone (eg "git://github.com/xwiki/xwiki-commons.git")
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of Git user
+     * @param token the Personal access token generated by the user
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/GitManager.java]: Maybe mention how to get that token. Hmmm what token is this about?

File: src/main/java/org/xwiki/git/internal/DefaultGitManager.java [hunk 4]
```diff
@@ -67,32 +69,42 @@
     @Inject
     private Logger logger;
 
-    @Override
-    public Repository getRepository(String repositoryURI, String localDirectoryName)
+    private Repository getGitRepository(String repositoryURI, String localDirectoryName,
+                                        String username, String token)
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/internal/DefaultGitManager.java]: For indentations and code style please see https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/

File: src/main/java/org/xwiki/git/internal/DefaultGitManager.java [hunk 5]
```diff
@@ -67,32 +69,42 @@
...
         Repository repository;
 
+        // Specify the local directory where the repository will be saved.
         File localGitDirectory = new File(this.environment.getPermanentDirectory(), "git");
         File localDirectory = new File(localGitDirectory, localDirectoryName);
         File gitDirectory = new File(localDirectory, ".git");
         this.logger.debug("Local Git repository is at [{}]", gitDirectory);
-        FileRepositoryBuilder builder = new FileRepositoryBuilder();
 
+        // Initialize builder for the Git environment.
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/internal/DefaultGitManager.java:83]: I think this comment is not really useful, is it? :)

**Comment by vmassol** [comment]
@Haxsen the PR is good. I have only small comments that I have put. Cool.

One thing though, the test has not been updated with a new test to verify that it works.

File: src/main/java/org/xwiki/git/GitManager.java (hunk 1 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManager.java]: Yeah I was wondering about this if it's either the version of the extension or the xwiki.
Updated.

File: src/main/java/org/xwiki/git/GitManager.java (hunk 3 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManager.java]: Updated.

**Comment by Haxsen** [comment]
@vmassol Yep I thought it might not have tests since the repository is private (and need that user's credentials).
I will update you with the tests.

**Comment by vmassol** [comment]
The test creates the git repo. I still don't know what a personal access token is for git. We're talking about git, not github, right?

**Comment by Haxsen** [comment]
@vmassol  Yes, not just for github. Personal access token is required to authenticate with the user ID. It also works with Two-factor authentication. See [Gitlab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) and [Github](https://developer.github.com/v3/auth/#working-with-two-factor-authentication) pages about it.
Forgot that there's also others like GitLab, my mistake.
Updated the javadoc.
Let me know if there are more mistakes.

File: src/main/java/org/xwiki/git/internal/DefaultGitManager.java (hunk 4 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/internal/DefaultGitManager.java]: Updated. Please confirm. (will update the others then)

**Comment by vmassol** [comment]
> Yes, not just for github. Personal access token is required to authenticate with the user ID. It also works with Two-factor authentication. See Gitlab and Github pages about it.

I don't understand it. This is a git module not a gitlab or github module. If I create a git repo locally, it should work. I admit that I don't know how git permissions work.

EDIT: I guess the concept of private git repo only exists for the HTTP protocol and there's a generic spec for securing a remote git repo. Would you have a link?

**Comment by vmassol** [comment]
Just checked UsernamePasswordCredentialsProvider at https://github.com/eclipse/jgit/blob/master/org.eclipse.jgit/src/org/eclipse/jgit/transport/UsernamePasswordCredentialsProvider.java and it refers to username and passwords, not token. So I have the feeling that the API you added is not correct and that the javadoc you added is even more wrong :)

Meaning that it'll work with any remote server that protects a git repo, whether it uses tokens, simple passwords, etc.

**Comment by Haxsen** [comment]
@vmassol 
On the [Github page](https://developer.github.com/v3/auth/#working-with-two-factor-authentication) I linked, it says:
"Deprecation Notice: GitHub will discontinue password authentication to the API."
If we use password authentication, it will deprecated.
So what do you suggest? Will follow your guide.

**Comment by vmassol** [comment]
@Haxsen I'll repeat it once more: we are NOT doing an API for GitHub or GitLab. We're writing a generic Git API. I just want you to understand what you're doing... you need to understand why your javadoc is just wrong. That was my original question, remember?  It's definitely not about what I want or don't want. I'm not the one writing the git spec!

**Comment by Haxsen** [comment]
@vmassol Ok, give me some time I will research more into this.
Will update you. Thanks.

**Comment by Haxsen** [comment]
@vmassol 
Ah I understand it now, I feel dumb for using token as param.
Now I have used `accessCode` which is more generic and can be password/oauth/token 😄 
Updated.
Let me know if I did it right or it has more problems.
I apologize for the frustration :/

File: src/main/java/org/xwiki/git/GitManager.java [hunk 6]
```diff
@@ -48,6 +49,25 @@
...
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of Git user
+     * @param accessCode the password or OAuth or personal access token that authenticates with the git user id.
+     * @return the cloned Repository instance
+     * @since 9.8
+     */
+    @Unstable
+    default Repository getPrivateRepository(String repositoryURI, String localDirectoryName, String username,
+            String accessCode)
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/GitManager.java]: just 4 spaces indentation. Could you fix all places please? Thx

**Comment by vmassol** [comment]
Looks good (except the indentation)! :)We're getting there

File: src/main/java/org/xwiki/git/GitManager.java (hunk 6 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManager.java]: Sure.

Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManager.java]: Done @vmassol . Can you confirm if I did right?

**Comment by Haxsen** [comment]
Fixing indentation, also: I have XWiki IntelliJ codestyle set up from the Java code style page. I press enter and it indents 8 spaces instead of 4.

**Comment by vmassol** [comment]
> Fixing indentation, also: I have XWiki IntelliJ codestyle set up from the Java code style page. I press enter and it indents 8 spaces instead of 4.

Weird. Maybe that intellij codestyle is wrong? Could you check? I looked at https://raw.githubusercontent.com/xwiki/xwiki-commons/master/xwiki-commons-tools/xwiki-commons-tool-verification-resources/src/main/resources/codestyle-idea.xml and I only see the number 4. No 8.

**Comment by Haxsen** [comment]
It's 4 on my end as well. I have restarted my IntelliJ several times. (Let me try Invalidate Caches and Restart).

**Comment by vmassol** [comment]
Applied, thanks!

**Comment by Haxsen** [comment]
You're welcome and thanks!

**Comment by vmassol** [comment]
Actually it's not good yet... You forgot to run the build with the quality profile. See https://ci.xwiki.org/view/Contrib/job/XWiki%20Contrib/job/api-git/job/master/36/console . You dropped the quality of the module, hence the build failure.

> [WARNING] Rule violated for bundle api-git: instructions covered ratio is 0.81, but expected minimum is 0.87

**Comment by Haxsen** [comment]
@vmassol 
I see, hm. Let me learn about this problem. I would try to reproduce this issue then fix. I think it might be due to tests, but let's see.
Thanks for letting me know.

**Comment by vmassol** [comment]
It is due to the tests covering less code than before since you added code but not tests :)

**Comment by Haxsen** [comment]
@vmassol Thanks for the hint. I am working on test now.

**Comment by Haxsen** [comment]
Hi @vmassol , I fixed indentation manually by putting 4 in codestyle settings (in IDEA) as it picked 8 (maybe IDEA bug).
I made tests and built on windows but windows didn't throw coverage errors. So I had to reproduce the issue in order to fix. I found that windows has many problems than linux in terms of maven building (and problems with java files, code styles etc. as well). So I had to install linux and test on it.
Just to mention, openJDK latest version fails to build (java crashes) so I used openJDK 8 instead.
PR sent for tests.
Thanks.

**Pull request merged**